### ISSUE TITLE
[Formatter] newLine + spaces can't be achieved

### ISSUE THREAD

**Issue by moayman** [eclipse/xtext-core#1069]
I am trying to indent a list based on its header and I can't achieve this using the current formatter.

Further explanation:

Grammar:
`grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals

generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"

DSL returns XTDSL:
	models+=Model*
;

Model returns XTModel:
	{XTModel} 'startList' '(' greetings+=Greeting*')'';';
	
Greeting returns XTGreeting:
	'list' ids_list=identifier_list ':' 'endlist' ';';

identifier_list returns XTIdentifierList:
	ids+=identifier (',' ids+=identifier)*
;

identifier returns XTIdentifier:
	{XTIdentifier} ID
;`

Formatter code:
`def dispatch void format(XTDSL xtdsl, extension IFormattableDocument document) {
		xtdsl.models.forEach[format]
	}

	def dispatch void format(XTModel xtmodel, extension IFormattableDocument document) {
		xtmodel.greetings.forEach[format; append[newLine]]
		xtmodel.regionFor.keyword("startList").append[oneSpace]
		xtmodel.regionFor.keyword("(").append[newLine]
		xtmodel.regionFor.keyword(")").prepend[noSpace]
		xtmodel.regionFor.keyword(";").prepend[noSpace]
		xtmodel.append[newLine]
		alignGreetingsOnOpenBracket(xtmodel.greetings, document)
	}

	def void alignGreetingsOnOpenBracket(EList<XTGreeting> greetings, extension IFormattableDocument document) {
		val precedingString = "startList ("
		for (greeting : greetings) {
			greeting.prepend[space = Strings.repeat(" ", precedingString .length); highPriority]
		}
	}`

After applying the formatter:
`startList (
list A : endlist;
list B : endlist;
list C : endlist;
);`

Expected output after applying the formatter:
`startList (
           list A : endlist;
           list B : endlist;
           list C : endlist;
);`

Basically what I am trying to achieve is having spaces before **list A : endlist;** with the width of **startList (** to have the list A, list B and list C all aligned with the opening bracket found after startList.

The formatter implementation does not allow something like this:
`var precedingString = "startList ("
greeting.prepend[**space = Strings.repeat(" ", precedingString.length);** newLine]`

If a newLine exists then spaces specified will always be ignored. The only way to have spaces with a newLine is to use indent which can be of whatever width specified by tabWidth EditorsPlugin preference.

I assume that the formatter should allow having spaces along with newLines or allow the specification of the indentation width as an override to the tabWidth preference if needed.

**Comment by tivervac** [comment]
We've added this feature in our product but it's not an easy fix and there are tons of edge cases.
The bulk of changes to require this to work are in `org.eclipse.xtext.formatting2.internal.WhitespaceReplacer` and `org.eclipse.xtext.formatting2.internal.HiddenRegionReplacer`

**Comment by moayman** [comment]
@tivervac I am sorry, but what product are you talking about?

**Comment by tivervac** [comment]
@moayman That would be [Sigasi Studio](https://www.sigasi.com/). However, our product isn't open source so I'm afraid I can't help you with that, if you'd like to customize your own product to support this feature you can always drop me a mail or you can pm me on the Xtext gitter

**Comment by moayman** [comment]
@tivervac Thanks a lot. I will be contacting you later about it.

**Comment by ArneDeutsch** [comment]
Your effect can be achieved without customizing other classes. Basically you would do the following:

1. Create a method `formatForIndention` that you call for each of your `Greeting` elements (probably you will add `index `and `count` as parameters to handle first and last element in special way).
2. Inside the method you calculate the needed indentation based on the parent element of your `Greeting`. This method might be a bit complex to handle all side cases, but in effect you have to calculate only the number of blancks you want to insert).
3. Call `setSpace` for the region. Something like this (you probably need to replace `featur(LIST_NAME)` with a method to get the `'list'` keyword that you want to prepend).
```
val nameColumn = nameColumnOfElement(variable.eContainer)
variable.regionFor.feature(LIST__NAME).prepend [
	space = Strings.repeat(" ", nameColumn(variable.eContainer) - nameColumn + 1)
]
```

**Comment by tivervac** [comment]
@moayman I just remembered I have a PR open where I quickly created a basic language that allows spaces to be added together with newlines. Note that this solution is not exactly what we're using in our product. That solution is a lot more complex, mostly to be able to deal with all kinds of comments mixed in to the code.

Anyway, here's the PR: https://github.com/eclipse/xtext-core/pull/199/files

**Comment by moayman** [comment]
@ArneDeutsch Well, I believe you meant 
```
space = "\n" + Strings.repeat(" ", nameColumn(variable.eContainer) - nameColumn + 1)
```
didn't you?
If so, it seems more like a workaround.

**Comment by cdietrich** [comment]
that is more or less what i proposed as workaround here: https://www.eclipse.org/forums/index.php?t=msg&th=1097735&goto=1803170&#msg_1803170

**Comment by moayman** [comment]
@tivervac I have to take my time looking at the PR. But umm, I am having serious issues with comments. The formatter is pretty weak when it comes to comments. so sad.

For example:

1. Comments doesn't follow indentation if they're at the beginning of an indented region.
2. A single line comment counts as a new line, for example lets say I do this `append[newLines(3,3,3)]` between two lines and there exists a single line comment between those lines the result would be:

```
line 1
// comment (counts as a line from the 3)

line 2
```
while I expect:
```
line 1
// comment


line 2
```

**Comment by moayman** [comment]
@cdietrich Yes exactly, this workaround certainly does the job. But I believe that the formatter could use some enhancements.

**Comment by tivervac** [comment]
Nice workaround, never thought of that actually.

@moayman I feel your pain regarding comments. I've definitely fixed that issues but I'll need to have a good look around our code to check which patch fixes this exact issue. As you can imagine, I've fixed tens of similar issues

**Comment by tivervac** [comment]
@moayman Got it, reported it before over [here](https://github.com/eclipse/xtext-xtend/issues/77)
